db/python/tables/family.py

```diff
@@ -167,14 +179,60 @@ async def get_family_external_ids_by_participant_ids(
     async def update_family(
         self,
         id_: int,
-        external_id: str | None = None,
+        external_ids: dict[str, str | None] | None = None,
         description: str | None = None,
         coded_phenotype: str | None = None,
     ) -> bool:
         """Update values for a family"""
-        values: Dict[str, Any] = {'audit_log_id': await self.audit_log_id()}
-        if external_id:
-            values['external_id'] = external_id
+        audit_log_id = await self.audit_log_id()
+        values: Dict[str, Any] = {'audit_log_id': audit_log_id}
+
+        if external_ids:
+            to_delete = [k.lower() for k, v in external_ids.items() if v is None]
+            to_update = {k.lower(): v for k, v in external_ids.items() if v is not None}
+
+            if to_delete:
+                # Set audit_log_id to this transaction before deleting the rows
+                _audit_update_query = """
+UPDATE family_external_id
+SET audit_log_id = :audit_log_id
+WHERE family_id = :id AND name IN :names
+                """
+                await self.connection.execute(
+                    _audit_update_query,
+                    {'id': id_, 'names': to_delete, 'audit_log_id': audit_log_id},
+                )
+
+                _delete_query = """
+DELETE FROM family_exernal_id
+WHERE family_id = :id AND name in :names
+                """
+                await self.connection.execute(
+                    _delete_query, {'id': id_, 'names': to_delete}
+                )
+
+            if to_update:
+                _query = 'SELECT project FROM family WHERE id = :id'
+                row = await self.connection.fetch_one(_query, {'id': id_})
+                project = row['project']
+
+                _update_query = """
+INSERT INTO family_external_id (project, family_id, name, external_id, audit_log_id)
+VALUES (:project, :id, :name, :external_id, :audit_log_id)
+ON DUPLICATE KEY UPDATE external_id = :external_id, audit_log_id = :audit_log_id
```

Comment:
I think this will update the audit log id even if the `external_id` isn't changed. It could be worth making that update conditional so that it only gets updated if it is different, ie.

```sql
ON DUPLICATE KEY UPDATE
    audit_log_id = IF(
        external_id = VALUES(external_id),
        audit_log_id, -- Return the existing value if unchanged
        VALUES(audit_log_id) -- Return the new value.
    ),
    external_id = IF(
        external_id = VALUES(external_id),
        external_id,
        VALUES(external_id)
    )
```

I haven't tested if this works, but I think should be ok in theory - what I'm not sure about is whether it'll still create a entry in the system versioned tables if the `IF` statement returns the exact existing value. Even if it does I think it'd still be good cos the audit_log_id wouldn't be updated.

Reply:
Alternatively could filter out unchanged external ids before doing the update

Reply:
After consideration, I'm going to call this a **feature** :smile:

When updating, you have the option of leaving items out of `external_ids` if you don't want them changed. So if you do specify `'foo': 'bar'` even though its value is already `'bar'`, I'm going to call it a feature that that causes this record to be bumped in the audit log, and `foo`'s value to be “freshened”.

Reply:
No worries, sounds fine. I think there's plenty of places where the audit log is a tad noisier than it could be so we'll probably need to implement some de-duplication if we ever want to display in the UI/API anyway